mypy/dmypy_server.py

```diff
@@ -212,6 +235,71 @@ def check(self, sources: List[mypy.build.BuildSource],
             response.update(self.last_manager.stats_summary())
         return response
 
+    def check_fine_grained(self, sources: List[mypy.build.BuildSource]) -> Dict[str, Any]:
+        """Check using fine-grained incremental mode."""
+        if not self.fine_grained_initialized:
+            return self.initialize_fine_grained(sources)
+        else:
+            return self.fine_grained_increment(sources)
+
+    def initialize_fine_grained(self, sources: List[mypy.build.BuildSource]) -> Dict[str, Any]:
+        try:
+            # TODO: alt_lib_path
+            result = mypy.build.build(sources=sources,
+                                      options=self.options)
+        except mypy.errors.CompileError as e:
+            # TODO: Handle blockers here
+            assert False, str('\n'.join(e.messages))
+        messages = result.errors
+        manager = result.manager
+        graph = result.graph
+        self.fine_grained_manager = mypy.server.update.FineGrainedBuildManager(manager, graph)
+        status = 1 if messages else 0
+        self.previous_messages = messages[:]
+        if messages:
+            messages.append('')
+        self.fine_grained_initialized = True
+        self.file_modified = {}  # type: Dict[str, float]
+        for source in sources:
+            assert source.path
+            self.file_modified[source.path] = os.stat(source.path).st_mtime
+        self.previous_sources = sources
+        return {'out': '\n'.join(messages), 'err': '', 'status': status}
+
+    def fine_grained_increment(self, sources: List[mypy.build.BuildSource]) -> Dict[str, Any]:
+        changed = self.find_changed(sources)
+        if not changed:
+            # Nothing changed -- just produce the same result as before.
+            messages = self.previous_messages
+        else:
+            messages = self.fine_grained_manager.update(changed)
+        status = 1 if messages else 0
+        self.previous_messages = messages[:]
+        if messages:
+            messages.append('')
+        self.previous_sources = sources
+        return {'out': '\n'.join(messages), 'err': '', 'status': status}
+
+    def find_changed(self, sources: List[mypy.build.BuildSource]) -> List[Tuple[str, str]]:
+        changed = []
+        for source in sources:
+            path = source.path
+            assert path
+            mtime = os.stat(path).st_mtime
+            if path not in self.file_modified or self.file_modified[path] != mtime:
+                self.file_modified[path] = mtime
+                changed.append((source.module, path))
+        modules = {source.module for source in sources}
+        omitted = [source for source in self.previous_sources if source.module not in modules]
+        for source in omitted:
+            path = source.path
+            assert path
+            if not os.path.isfile(path):
```

Comment:
What cases other than the file being removed would cause being omitted?

Reply:
The file could be removed from the root sources, but it could still continue to exist on the file system. I'll add a comment.